src/toil/job.py

```diff
@@ -3758,6 +3826,181 @@ def getUserScript(self):
+            if imported is not None:
+                path_to_fileid[file] = imported
+        return path_to_fileid
+
+    def run(self, file_store: AbstractFileStore) -> Promised[Dict[str, FileID]]:
+        """
+        Import the workflow inputs and then create and run the workflow.
+        :return: Promise of workflow outputs
+        """
+        try:
+            return self.import_files(self.filenames, file_store.jobStore)
+        except InsufficientSystemResources as e:
+            if e.resource == "disk" and e.requested < self.disk_size:
+                non_streaming_import = WorkerImportJob(
+                    self.filenames, self.disk_size, streaming=False
+                )
+                self.addChild(non_streaming_import)
+                return non_streaming_import.rv()
+            else:
+                raise
+
+
+class ImportsJob(Job):
+    """
+    Job to organize and delegate files to individual WorkerImportJobs.
+
+    For the CWL/WDL runners, this is only used when runImportsOnWorkers is enabled
+    """
+
+    def __init__(
+        self,
+        file_to_data: Dict[str, FileMetadata],
+        import_workers_threshold: ParseableIndivisibleResource,
+        **kwargs: Any,
+    ):
+        """
+        Job to take the inputs from the WDL workflow and import them on a worker instead of a leader. Assumes all local and cloud files are accessible.
+
+        This class is only used when runImportsOnWorkers is enabled.
+        """
+        super().__init__(local=True, **kwargs)
+        self._file_to_data = file_to_data
+        self._import_workers_threshold = import_workers_threshold
+
+    def run(
+        self, file_store: AbstractFileStore
+    ) -> Tuple[Promised[Dict[str, FileID]], Dict[str, FileMetadata]]:
+        """
+        Import the workflow inputs and then create and run the workflow.
+        :return: Tuple of a mapping from the candidate uri to the file id and a mapping of the source filenames to its metadata. The candidate uri is a field in the file metadata
+        """
+        import_workers_threshold = self._import_workers_threshold
+        file_to_data = self._file_to_data
+        # Run WDL imports on a worker instead
+
+        individual_files = []
+        batch_files = []
+        # Separate out files that will be imported individually and in a batch
+        for filename, (candidate_uri, parent_id, size) in file_to_data.items():
+            if size < import_workers_threshold:
+                batch_files.append(filename)
+            else:
+                individual_files.append(filename)
+
+        # Create individual import jobs
+        import_jobs = []
+        for file in individual_files:
+            filesize = file_to_data[file][2]
+            candidate_uri = file_to_data[file][0]
+            import_jobs.append(WorkerImportJob([candidate_uri], disk_size=filesize))
+
+        # create multiple batch jobs if necessary to not exceed the threshold
+        per_batch_files = []
```

Comment:
This sounds like it would hold a list of lists: for each batch, a list of the files in that batch. The comment above it also suggests it might hold "multiple batch jobs". But actually it holds the current batch under construction.

I would initialize the current thing under construction *after* the list that is going to accumulate all the things constructed.